Lib/gzip.py

```diff
@@ -430,29 +430,42 @@ def _read_gzip_header(fp):
 
     if magic != b'\037\213':
         raise BadGzipFile('Not a gzipped file (%r)' % magic)
-
-    (method, flag, last_mtime) = struct.unpack("<BBIxx", _read_exact(fp, 8))
+    header_buffer = io.BytesIO()
```

Comment:
Try to use a `bytearray` instead of `io.BytesIO`. Would it be faster? In what cases?

Reply:
Performance improvement was necessary (see bug report). So I made it better the following way:

- When the data is read from `fp` a bytes object is created. Instead of passing it anonymously to the functions that performs the check it is saved as a local variable.
- The FHCRC flag is checked and the result is saved as a variable, so the check runs only once.
- When `fhcrc` is truthy, a `header_parts` list is created. 
- Every time  `fp` is read and `fhcrc` is truthy the already created bytes objects are appended to the `header_parts` list. List appending has a very low overhead. It simply expands an array of PyObject pointers. Pointers that we already have as the named bytes objects.
- When the header should be checked, a `b"".join(header_parts)` is called to recreate the header.

This means for the most common use cases (no flags or only FNAME set) that only a few extra truthy checks are performed. This comes at a very low cost of 3% (see the bug report for results).

This is worth it since it enables correct truncation checking for FNAME and FCOMMENT as well as enabling checking of the header CRC when FHCRC is set.